matchzoo/preprocessors/units/tokenize.py

```diff
@@ -15,3 +17,104 @@ def transform(self, input_: str) -> list:
         :return tokens: tokenized tokens as a list.
         """
         return nltk.word_tokenize(input_)
+
+
+class ChineseTokenize(Unit):
+    """Process unit for text containing Chinese tokens."""
+
+    def transform(self, input_: str) -> str:
+        """
+        Process input data from raw terms to processed text.
+
+        :param input_: raw textual input.
+
+        :return output: text with a blank between adjacent Chinese tokens.
+        """
+        output = []
+        for char in input_:
+            cp = ord(char)
+            if is_chinese_char(cp):
+                output.append(" ")
+                output.append(char)
+                output.append(" ")
+            else:
+                output.append(char)
+        return "".join(output)
+
+
+class BasicTokenize(Unit):
+    """Process unit for text tokenization."""
+
+    def transform(self, input_: str) -> list:
+        """
+        Process input data from raw terms to list of tokens.
+
+        :param input_: raw textual input.
+
+        :return tokens: tokenized tokens as a list.
+        """
+        orig_tokens = whitespace_tokenize(input_)
+        split_tokens = []
+        for token in orig_tokens:
+            split_tokens.extend(run_split_on_punc(token))
+
+        output_tokens = whitespace_tokenize(" ".join(split_tokens))
+        return output_tokens
+
+
+class WordPieceTokenize(Unit):
+    """Process unit for text tokenization."""
+
+    def __init__(self, vocab: dict, max_input_chars_per_word: int = 200):
+        """Initialization."""
+        self.vocab = vocab
+        self.unk_token = '[UNK]'
+        self.max_input_chars_per_word = max_input_chars_per_word
+
+    def transform(self, input_: list) -> list:
+        """
+        Tokenizes a piece of text into its word pieces.
+
+        This uses a greedy longest-match-first algorithm to perform
+         tokenization using the given vocabulary.
+
+        For example:
+          input = "unaffable"
+          output = ["un", "##aff", "##able"]
+
+        :param input_: token list.
+
+        :return tokens: A list of wordpiece tokens.
+        """
+        output_tokens = []
+        for token in input_:
+            chars = list(token)
+            if len(chars) > self.max_input_chars_per_word:
+                output_tokens.append(self.unk_token)
+                continue
+
+            is_bad = False
+            start = 0
+            sub_tokens = []
+            while start < len(chars):
+                end = len(chars)
```

Comment:
why put this line inside of the while loop?

Reply:
At line 110, if the prefix is not matched in vocab, the end index shrinks by step to find the longest prefix and tokenize.

Reply:
but `len(chars)` never changed, right?

Reply:
Yes, I set a variable to store it.